### PR TITLE
NXP-30945: Import compound documents previews

### PR DESCRIPTION
**Pull request by NourNuxeo**

### CODE REVIEW

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 1]
```diff
@@ -138,6 +140,10 @@ protected void createEntry(CoreSession session, DocumentModel compoundDoc, ZipFi
                 throw new NuxeoException(message, e);
             }
         }
+        if (FilenameUtils.removeExtension(entry.getName()).equals("preview")) {
+            compoundDoc.setPropertyValue("cp:preview", doc.getId());
+            session.saveDocument(compoundDoc);
+        }
```
Comment (troger) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: This should go in the `else` part, right? I don't think we want to set a directory as the preview 😬

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        // fallback to default factory
+        return getDefaultThumbnail(doc);
+    }
+
+    public Blob getThumbnailFromDocId(String docId, CoreSession session) {
+        if (StringUtils.isNotEmpty(docId)) {
+            ThumbnailService thumbnailService = Framework.getService(ThumbnailService.class);
+            DocumentModel previewDoc = session.getDocument(new IdRef(docId));
```
Comment (troger) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java]:
```suggestion
            DocumentModel doc = session.getDocument(new IdRef(docId));
```
Let's call it `doc`, it's not always a preview doc.

File: nuxeo-compound-documents/src/main/resources/OSGI-INF/thumbnailfactory-contrib.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<component name="org.nuxeo.compound.documents.thumbnail">
+  <require>org.nuxeo.ecm.platform.thumbnail.service.thumbnailfactory</require>
+  <extension point="thumbnailFactory" target="org.nuxeo.ecm.core.api.thumbnail.ThumbnailService">
+    <thumbnailFactory facet="CompoundDocument" factoryClass="org.nuxeo.compound.documents.CompoundDocumentThumbnailFactory" name="compoundDocumentsThumbnailFactory" />
```
Comment (troger) [inline nuxeo-compound-documents/src/main/resources/OSGI-INF/thumbnailfactory-contrib.xml]:
```suggestion
    <thumbnailFactory facet="CompoundDocument" factoryClass="org.nuxeo.compound.documents.CompoundDocumentThumbnailFactory" name="compoundDocumentThumbnailFactory" />
```

File: nuxeo-compound-documents/src/main/resources/schemas/compound.xsd [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           xmlns:nxs="http://www.nuxeo.org/ecm/project/schemas/compound"
+           xmlns:nxsv="http://www.nuxeo.org/ecm/schemas/core/validation/"
+           xmlns:ref="http://www.nuxeo.org/ecm/schemas/core/external-references/"
+           targetNamespace="http://www.nuxeo.org/ecm/project/schemas/compound">
+  <xs:element name="preview">
+    <xs:simpleType>
+      <xs:restriction base="xs:string" ref:resolver="documentResolver" ref:store="id"/>
```
Comment (troger) [inline nuxeo-compound-documents/src/main/resources/schemas/compound.xsd]:
```suggestion
      <xs:restriction base="xs:string" ref:resolver="documentResolver" ref:store="id" />
```

File: nuxeo-compound-documents/src/main/resources/schemas/compound.xsd [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    </xs:simpleType>
+  </xs:element>
+  <xs:element name="thumbnail">
+    <xs:simpleType>
+      <xs:restriction base="xs:string" ref:resolver="documentResolver" ref:store="id"/>
```
Comment (troger) [inline nuxeo-compound-documents/src/main/resources/schemas/compound.xsd]:
```suggestion
      <xs:restriction base="xs:string" ref:resolver="documentResolver" ref:store="id" />
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java [hunk 6]
```diff
@@ -70,6 +79,42 @@ public void testCreateCompoundDocument() throws IOException {
...
+    public void testCreateCompoundDocumentWithPreview() throws IOException {
+        Blob blob = getPreviewTestArchive();
+        FileImporterContext context = FileImporterContext.builder(session, blob, "/").build();
+
+        DocumentModel doc = fileManager.createOrUpdateDocument(context);
+
+        doc = session.getDocument(doc.getRef());
+        var previewId = doc.getPropertyValue("cp:preview");
+        var preview = session.getDocument(new PathRef("/" + doc.getName() + "/preview.png"));
+        assertTrue(preview.getName().equals("preview.png"));
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java]:
```suggestion
        assertEquals("preview.png", preview.getName());
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java [hunk 7]
```diff
@@ -70,6 +79,42 @@ public void testCreateCompoundDocument() throws IOException {
...
+    @Test
+    public void testCreateCompoundDocumentWithPreview() throws IOException {
+        Blob blob = getPreviewTestArchive();
+        FileImporterContext context = FileImporterContext.builder(session, blob, "/").build();
+
+        DocumentModel doc = fileManager.createOrUpdateDocument(context);
+
+        doc = session.getDocument(doc.getRef());
+        var previewId = doc.getPropertyValue("cp:preview");
+        var preview = session.getDocument(new PathRef("/" + doc.getName() + "/preview.png"));
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java]:
```suggestion
        var preview = session.getDocument(new PathRef(doc.getPathAsString(), "preview.png"));
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java [hunk 8]
```diff
@@ -70,6 +79,42 @@ public void testCreateCompoundDocument() throws IOException {
...
+        txFeature.nextTransaction();
+
+        var thumbnailId = doc.getPropertyValue("cp:thumbnail");
+        assertNull(thumbnailId);
+        var previewId = doc.getPropertyValue("cp:preview");
+        assertNotNull(previewId);
+        var preview = session.getDocument(new PathRef("/" + doc.getName() + "/preview.png"));
+        assertNotNull(preview);
+        assertEquals(preview.getId(), previewId);
+        ThumbnailService thumbnailService = Framework.getService(ThumbnailService.class);
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java]: Inject it in the test class (like other services).

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/CompoundDocumentsFeature.java [hunk 9]
```diff
@@ -26,7 +26,12 @@
 @Deploy("org.nuxeo.ecm.platform.filemanager")
 @Deploy("org.nuxeo.ecm.platform.video:OSGI-INF/core-types-contrib.xml")
 @Deploy("org.nuxeo.ecm.platform.audio.core:OSGI-INF/core-types-contrib.xml")
-@Deploy("org.nuxeo.ecm.platform.picture.core:OSGI-INF/picture-schemas-contrib.xml")
+@Deploy("org.nuxeo.ecm.platform.picture.core")
+@Deploy("org.nuxeo.ecm.platform.convert")
+@Deploy("org.nuxeo.ecm.platform.thumbnail")
+@Deploy("org.nuxeo.ecm.platform.rendition.core")
+@Deploy("org.nuxeo.ecm.platform.io.core")
+@Deploy("org.nuxeo.ecm.platform.webapp.types")
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/CompoundDocumentsFeature.java]:
```suggestion
@Deploy("org.nuxeo.ecm.platform.picture.core")
@Deploy("org.nuxeo.ecm.platform.thumbnail")
```
I would keep it simple as long as tests pass.

**Review by troger — COMMENTED** [review]
Nice!

Both tests `testCreateCompoundDocumentWithPreview` and `testCompoundDocumentWithPreviewAsThumbnail` can be merged in only one test.

I am wondering about the thumbnail stuff (`cp:thumbnail`), what's used for? It's never computed nor set. I would remove any reference for now as we don't have any use case.

I would put a lighter image as preview, we just need a 4 pixels one :)

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java (hunk 2 above)
Comment (NourNuxeo) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java]: While I'm at it, would the following be a var keyword overdose :
```
public class CompoundDocumentThumbnailFactory extends ThumbnailDocumentFactory {

    @Override
    public Blob getThumbnail(DocumentModel doc, CoreSession session) {

        // give priority to thumbnail property
        if (doc.hasFacet(THUMBNAIL_FACET)) {
            var blob = (Blob) doc.getPropertyValue(THUMBNAIL_PROPERTY_NAME);
            if (blob != null) {
                return blob;
            }
        }

        // then from the preview document
        var previewDocId = (String) doc.getPropertyValue("cp:preview");
        Blob thumbnailBlob = getThumbnailFromDocId(previewDocId, session);
        if (thumbnailBlob != null) {
            return thumbnailBlob;
        }

        // fallback to default factory
        return getDefaultThumbnail(doc);
    }

    public Blob getThumbnailFromDocId(String docId, CoreSession session) {
        if (StringUtils.isNotEmpty(docId)) {
            var thumbnailService = Framework.getService(ThumbnailService.class);
            var doc = session.getDocument(new IdRef(docId));
            return thumbnailService.getThumbnail(doc, session);
        }
        return null;
    }

}
```

Comment (troger) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java]:
```
var thumbnailBlob = getThumbnailFromDocId(previewDocId, session);
```
😬

File: nuxeo-compound-documents/pom.xml [hunk 10]
```diff
@@ -12,6 +12,14 @@
...
 
   <dependencies>
+    <dependency>
+      <groupId>org.nuxeo.ecm.platform</groupId>
+      <artifactId>nuxeo-platform-rendition-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.ecm.platform</groupId>
+      <artifactId>nuxeo-platform-io-core</artifactId>
+    </dependency>
```
Comment (troger) [inline nuxeo-compound-documents/pom.xml]: The pom probably needs to be cleaned up as some bundles were removed from the test feature.

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java (hunk 2 above)
Comment (NourNuxeo) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentThumbnailFactory.java]: 😅 got it !

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 11]
```diff
@@ -117,16 +117,18 @@ protected void createEntry(CoreSession session, DocumentModel compoundDoc, ZipFi
         Path entryParentPath = entryPath.removeLastSegments(1);
         String parentDocPath = compoundDoc.getPath().append(entryParentPath).toString();
         String entryDocName = entryPath.lastSegment();
+        DocumentModel doc;
```
Comment (troger) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: Not needed anymore.

File: nuxeo-compound-documents/src/main/resources/schemas/compound.xsd [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           xmlns:nxs="http://www.nuxeo.org/ecm/project/schemas/compound"
+           xmlns:nxsv="http://www.nuxeo.org/ecm/schemas/core/validation/"
```
Comment (troger) [inline nuxeo-compound-documents/src/main/resources/schemas/compound.xsd]: Not needed I think (no use of `nxsv` in the schema definition).

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java (hunk 11 above)
Comment (NourNuxeo) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: I would still need it in the else as it is accessed in the try and the last if. Do I move it down?

Comment (troger) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: Just use `var doc = ...` in both if/else, it's not used outside the if/else.

Comment (NourNuxeo) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: Pushed ! 👍

**Review by guirenard — APPROVED** [review]

**Review by troger — APPROVED** [review]

**Pull request merged**